### PR TITLE
Tagging on grasp

### PR DESCRIPTION
**Pull request by hessikaveh**
Adding the tagging feature,
Opening a pull request on the first working version, to fix styling and other issues.
Issues:
*The undo and redo is not implemented 
*scroll behavior on the Tags column of the table 

ezgif-2-24ad8b319c35
can be tested here:
https://pdmv-dev-proxy.web.cern.ch/samples/existing?name=RunIISummer20UL17*GEN&pwg=JME

### CODE REVIEW

File: update_scripts/update_existing_samples.py [hunk 1]
```diff
@@ -165,10 +165,11 @@ def process_request(self, campaign_uid, request):
             root_request = self.get_request_if_exists(root_request_prepid)
             miniaod_request = self.get_request_if_exists(miniaod_prepid)
             nanoaod_request = self.get_request_if_exists(nanoaod_prepid)
-
+            other_tags = sorted_join(request['tags'])
```
Comment (justinasr) [inline update_scripts/update_existing_samples.py:168]: This is completely missing the bidirectionality of GrASP and McM updates. `other_tags = sorted_json(request['tags'])` will always overwrite any changes in GrASP with tags in McM. Check the `process_interested_pwgs_update` method.

File: update_scripts/update_existing_samples.py [hunk 2]
```diff
@@ -165,10 +165,11 @@ def process_request(self, campaign_uid, request):
...
             # Take interested PWGs from NanoAOD if it exists
             # If not, then MiniAOD request if it exists
             # If it does not exist, use root request
+
```
Comment (justinasr) [inline update_scripts/update_existing_samples.py:172]: Redundant empty line.

File: frontend/src/router/index.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ import UserTags from '@/components/UserTags'
 import UserTagsEdit from '@/components/UserTagsEdit'
 import History from '@/components/History'
 import qs from 'qs';
+import axios from 'axios'
```
Comment (justinasr) [inline frontend/src/router/index.js:10]: I don't see why any of the changes in this file are needed?

File: frontend/src/components/ExistingSamples.vue [hunk 4]
```diff
@@ -181,6 +193,14 @@
       <span style="color: var(--v-anchor-base); cursor: pointer" @click="openPmpSelectedEntries()">pMp Historical</span>
     </footer>
 
+    <footer v-if="selectAllChecked || selectAllIndeterminate">
```
Comment (justinasr) [inline frontend/src/components/ExistingSamples.vue:196]: Why are there two footers in the page? The second one completely covers the first one and "Add PWG" and "Remove PWG" buttons are no longer there.

File: api/existing_samples_api.py [hunk 5]
```diff
@@ -294,26 +295,33 @@ def post(self):
                         raise Exception('"%s" is not a valid PWG' % pwg)
 
                 interested_pwgs = sorted_join(interested_pwgs)
+                other_tags = clean_split(entry['tags'])
```
Comment (justinasr) [inline api/existing_samples_api.py:298]: These tags are _other_ given what reference, i.e. what are _the main_ tags then?

File: frontend/src/components/ExistingSamples.vue [hunk 6]
```diff
@@ -69,6 +69,14 @@
       </select>
       <b>physics working group as interested PWG in samples</b>
     </div>
+    <div v-if="campaign.entries" class="align-center ma-2">
+      <b>I am going to add or remove</b>
+      <select v-model="selectedGrASPTag" class="ml-2 mr-2">
+        <option selected value=''></option>
+        <option v-for="tag in userTags" :key="tag.name" :value="tag.name">{{tag.name}}</option>
+      </select>
+      <b>GrASP Custom Tag in samples</b>
```
Comment (justinasr) [inline frontend/src/components/ExistingSamples.vue:78]: I'm not sure about "GrASP Custom Tag", why not just "tag"?

File: frontend/src/components/ExistingSamples.vue [hunk 7]
```diff
@@ -166,6 +175,9 @@
             <span class="remove-pwg-link" v-if="entry.interested_pwgs.includes(selectedPWG)" @click="removePWGFromEntriesAction(selectedPWG, [entry])">Remove {{selectedPWG}}</span>
           </template>
         </td>
+        <td style="max-width: 100px; min-height: 50px; min-width: 50px; overflow: auto">
+          {{entry.tags}}
```
Comment (justinasr) [inline frontend/src/components/ExistingSamples.vue:179]: Would it be possible to add the "Add tag" and "Remove tag" buttons like in interested PWGs cell?

**Review by justinasr — CHANGES_REQUESTED** [review]
There is another issue, not with this PR, but with GrASP itself. Whenever stuff is tagged, users probably immediately expect it to appear in the "User tagged samples" tables. Unfortunately we have two separate tables in DB for "existing" and "user tagged" samples, maybe it would be possible to somehow merge them and have only one.

**Comment by hessikaveh** [comment]
> There is another issue, not with this PR, but with GrASP itself. Whenever stuff is tagged, users probably immediately expect it to appear in the "User tagged samples" tables. Unfortunately we have two separate tables in DB for "existing" and "user tagged" samples, maybe it would be possible to somehow merge them and have only one.

In principle, we can just drop the user_tags and keep the existing_campaigns table selecting on tag as the query for the user_tag page or join two queries from the two separate tables of user_tag entries and existing_campaigns? What do you think?
I think I covered the rest of the issues!

File: update_scripts/update_user_samples.py [hunk 8]
```diff
@@ -26,7 +27,9 @@ def update(self):
         and process chained requests that these requests are members of
         """
         # Iterate through tags and add or update requests
-        table, entries_table = self.get_table_names()
+        table, entries_table, campaings_table = self.get_table_names_()
+        campaign_names = query(self.conn, campaings_table, ['name'])
+        campaign_names = [v.split('*')[0] for d in campaign_names for k, v in d.items()]
```
Comment (justinasr) [inline update_scripts/update_user_samples.py:32]: Hm, this makes an assumption that all campaigns (GEN, Mini, Nano) have identical substrings in the front. While it is true based on the PdmV policy, it's a risky assumption to take.

File: update_scripts/update_user_samples.py [hunk 9]
```diff
@@ -14,8 +15,8 @@ def __init__(self, db_connection, dev=True):
         ExistingSamplesUpdater.__init__(self, db_connection, dev)
         self.updated_chained_requests = None
 
-    def get_table_names(self):
-        return 'user_tags', 'user_tag_entries'
+    def get_table_names_(self):
```
[the remark this answers is not in the document]
Comment (justinasr) [reply update_scripts/update_user_samples.py:18]: Well, `get_table_names` is used in multiple places, including `get_existing_sample` and `insert_or_update` methods. If we are changing the whole structure so there would be only one table, then this method becomes irrelevant and probably only the `update` method should be overwritten. Maybe not even the whole method, but just the `requests = self.mcm.get('requests', 'tags=%s' % (tag_name))` part - essentially now the only thing that is different between existing samples and user samples is query done to McM to fetch relevant requests.

[the remark this answers is not in the document]
Comment (justinasr) [reply update_scripts/update_user_samples.py:18]: Well, `get_table_names` is used in multiple places, including `get_existing_sample` and `insert_or_update` methods. If we are changing the whole structure so there would be only one table, then this method becomes irrelevant and probably only the `update` method should be overwritten. Maybe not even the whole method, but just the `requests = self.mcm.get('requests', 'tags=%s' % (tag_name))` part - essentially now the only thing that is different between existing samples and user samples is query done to McM to fetch relevant requests.

**Pull request merged**